File: Contest/MinimumTimetoCompleteTrips.java
class Solution {
	// TC : O(nlogn)
	// SC: O(1)
	public long minimumTime(int[] time, int totalTrips) {


		long left = 1;
		long min = time[0];
		for(int t : time){
			min = Math.min(min, t);
		}
		long right = totalTrips * min;

		while(left < right){
			long mid = (right - left) / 2 + left;

			//How many trip we can do
			long total = totalTripInGivenTime(time, mid);
			//We cannot reach totalTrips, means mid is too small
			if(total < totalTrips){
				left = mid + 1;
			}else{
				//We can reach totalTrips, but we want right to be as small as possible
				//Therefore, we don't need to + 1
				right = mid;
			}
		}

		return left;
	}

	private long totalTripInGivenTime(int[] time, long givenTime){
		long totalTrips = 0l;
		for(int tripTime : time){
			totalTrips += ((long)givenTime) / tripTime;
		}
		return totalTrips;
	}
}

File: Leetcode2023/March2023/C++/Minimum Time To Complete Trips.cpp
class Solution {
public:
    long long calTime(vector<int>& time,long long curr_time)
    {
        long long res=0;
        for(auto& i:time)
        {
             res+=curr_time/i;
        }
        return res;
    }


    long long minimumTime(vector<int>& time, int totalTrips) {
        long double  l=1,r=(long double)(*max_element(begin(time),end(time))) * (long double)totalTrips;
        long long trip=0;
        while(l<r)
        {
            long double  mid =l+(r-l)/2;
            trip=calTime(time,mid);
            if(trip>=totalTrips) r=mid;
            else l=mid+1;
        }
        return (long long)l;
    }
};